### DISCUSSION TITLE
How to convert all database Guids stored as binary to text

### DISCUSSION THREAD

**Discussion by alexey-leonovich** [linq2db/linq2db#4599] [Q&A] [answered]
Hello.
I have an existing database where all Guids are stored as binary (Linq2Db default setting). I need to convert them all to text.
I can do that with `db.Execute()` for the [following SQL](https://github.com/dotnet/efcore/issues/15078#issuecomment-475784385):
```
UPDATE MyTable
SET GuidColumn = hex(substr(GuidColumn, 4, 1)) ||
                 hex(substr(GuidColumn, 3, 1)) ||
                 hex(substr(GuidColumn, 2, 1)) ||
                 hex(substr(GuidColumn, 1, 1)) || '-' ||
                 hex(substr(GuidColumn, 6, 1)) ||
                 hex(substr(GuidColumn, 5, 1)) || '-' ||
                 hex(substr(GuidColumn, 8, 1)) ||
                 hex(substr(GuidColumn, 7, 1)) || '-' ||
                 hex(substr(GuidColumn, 9, 2)) || '-' ||
                 hex(substr(GuidColumn, 11, 6))
WHERE typeof(GuidColumn) == 'blob';
```
But is there a higher level approach (type safe and error proof) with Linq2Db? Maybe some value converter or something?

**Comment by alexey-leonovich** [comment, marked as answer]
Hmm, looks like linq2db is smart enough to correctly read blob Guid even if the field is annotated as `(DbType = "TEXT")`:
```
using LinqToDB.Data;
using LinqToDB.DataProvider.SQLite;
using System;
using System.Linq;
using LinqToDB;
using LinqToDB.Mapping;
using NUnit.Framework;

namespace SQLiteGuidText
{
    [Table(Name = "Guids")] public class GuidBlob
    {
        [Column] public Guid? Id { get; set; }
    }

    [Table(Name = "Guids")] public class GuidText
    {
        [Column(DbType = "TEXT")] public Guid? Id { get; set; }
    }
    internal class Program
    {
        static void Main(string[] args)
        {
            var testGuid = Guid.Parse("4FC51C1F-2869-4FEB-B385-F6B40D9CD0DF");

            using (var conn = new DataConnection(SQLiteTools.GetDataProvider(ProviderName.SQLiteClassic),"Data Source=blob.db"))
            {
                SQLiteTools.CreateDatabase("blob.db", extension: ".db");
                conn.CreateTable<GuidBlob>();
                conn.Insert(new GuidBlob { Id = testGuid });

                var existingBlobGuid = conn.GetTable<GuidText>().First().Id;
                Assert.That(existingBlobGuid, Is.EqualTo(testGuid));
            }
        }
    }
}
```
@MaceWindu is that "heuristics" intended? So I can simply change my Guid annotations from blob to text and then just read my existing database and save it to another file (and all data will be migrated but Guids will be now saved as text)? And I don't need SQL or a converter?

**Reply by MaceWindu** [reply]
That's expected.

Type information on column mapping is used for:
- create table statements
- proper typing of parameters/literals
- and some other places where we need type information

When we read data, this information is not usually used as type information for read column provided by provider (database) and we cannot ignore it and try to read uniqueidentifier data as text (in this case). We read it as guid and then apply type converters, registered in mapping schema between database type and property type
